src/problems.jl

```diff
@@ -19,18 +19,18 @@ Solution(x::Array{T, 1}, status::Symbol, optval::T) where {T} =
 Solution(x::Array{T, 1}, y::Array{T, 1}, status::Symbol, optval::T) where {T} =
     Solution(x, y, status, optval, true)
 
-mutable struct Problem
+mutable struct Problem{T<:Real}
     head::Symbol
     objective::AbstractExpr
     constraints::Array{Constraint}
     status::Symbol
-    optval::Float64OrNothing
+    optval::Union{Real,Nothing}
```

Comment:
```suggestion
    optval::Union{T,Nothing}
```
? Otherwise the type parameter isn't actually used.

Reply:
Hmm, I might not be using the type parameters correctly here. I don't think `optval` should necessarily be of type `T`, since if the solver returns a Float64, it would be better to not have it converted to e.g. a `BigFloat` in case `T = BigFloat` (since then you might not know you only got back a Float64 from the solver). The reason I want to add the type parameter is for line 112 of the proposed version,
```julia
function conic_problem(p::Problem{T}) where {T}
```
That is, the type parameter is to pass the type to `conic_problem` so that it can use it when constructing `A`, `b`, and `c` in the body of `conic_problem`:
```julia
c = spzeros(T, var_size, 1)
A = spzeros(T, constr_size, var_size)
b = spzeros(T, constr_size, 1)
```
But since it isn't used to actually parametrize fields of the struct, maybe it should just be another field?

The reason I think it makes sense as a type parameter is that in an ideal world we might have
```julia
mutable struct Problem{T<:Real}
    head::Symbol
    objective::AbstractExpr{T}
    constraints::Array{Constraint{T}}
    status::Status
    optval::Union{Real,Nothing}
    model::Union{MOIModel{T}, Nothing}
    solution::Solution
    ... inner constructor
end
```
where an `AbstractExpr{T}` is an object that has only children of type `AbstractExpr{T}` or `Constant{T}` (or is itself a variable or a `Constant{T}`). (Actually, to support complex numbers we would want to allow `Constant{Complex{T}}` or `Constant{T}`). But right now, we allow `Constant`'s of various types to be mixed together in one problem, so that would be a bigger change. I'm up for making that bigger change though if it's better.